### DISCUSSION TITLE
Issues Adding Git LFS Repo as a Submodule: Not Appearing in .gitmodules

### DISCUSSION THREAD

**Discussion by vladimirzb** [git-lfs/git-lfs#5844] [Q&A] [answered]
I'm trying to add a Git repository that uses Git LFS as a submodule in my project. When I run `git submodule add <url_of_LFS_repo>`, the command completes without errors, but the submodule doesn't appear in the `.gitmodules` file, and there are no changes listed in `git status`.

What could be causing the Git LFS repository not to be added as a submodule, and how can I resolve this?

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.  I wasn't able to reproduce the situation you describe with a small test:
```
$ git init test
$ cd test
$ echo foo >foo.txt
$ git add foo.txt
$ git commit -m foo

$ git submodule add https://github.com/chrisd8088/lfs-demo demo
$ git status
On branch main
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   .gitmodules
	new file:   demo

$ cat .gitmodules 
[submodule "demo"]
	path = demo
	url = https://github.com/chrisd8088/lfs-demo
```

Can you try something like that as well?  Also, what versions of Git and Git LFS are you using, and on what platform?  If you supply the output of `git lfs env` and maybe also `git config -l` (with any private information redacted), that might help us identify what's going on.

**Reply by vladimirzb** [reply]
Hi! Thanks for the quick reply, @chrisd8088. I’ve recorded a video to demonstrate the issue I’m facing.

1. First, I clone the repository where I intend to use submodules.
2. I add a submodule from a repository, which works fine and gets correctly added.
   ```bash
   git submodule add https://github.com/vladimirzb/Cs.Computacion.git
   ```
3. Then, I add another submodule from a repo that uses Git LFS. It downloads the content, but it doesn’t get added as a submodule.
   ```bash
   git submodule add https://github.com/vladimirzb/SecundariaORT.git
   ```

https://github.com/user-attachments/assets/bd99e88f-a5b5-4807-9f1e-672b07cc4930


Here are the outputs from `git lfs env`:

```bash
~/D/G/Todos-Mis-Apuntes   ~…  git lfs env                           Wed 14 Aug 2024 03:53:12 AM -03
git-lfs/3.4.1 (GitHub; linux amd64; go 1.22.2)
git version 2.43.0

Endpoint=https://github.com/vladimirzb/Todos-Mis-Apuntes.git/info/lfs (auth=none)
LocalWorkingDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes
LocalGitDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes/.git
LocalGitStorageDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes/.git
LocalMediaDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes/.git/lfs/objects
LocalReferenceDirs=
TempDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/home/vladi/Documents/GitHub/Todos-Mis-Apuntes/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_ASKPASS=/usr/share/code/resources/app/extensions/git/dist/askpass.sh
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

And here are the outputs from `git config -l`:

```bash
~/D/G/Todos-Mis-Apuntes   ~…  git config -l                         Wed 14 Aug 2024 03:53:14 AM -03
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
user.name=Vladimir Z.
user.email=****@users.noreply.github.com
core.repositoryformatversion=0
core.filemode=true
```

Do you have any insights into why this might be happening? Thanks!

**Reply by vladimirzb** [reply]
If you need any more logs or information, just let me know! 😊 @chrisd8088

**Reply by chrisd8088** [reply]
Intruiging!  I've tried to reproduce this but so far haven't been able to do so.  (I'm using Git LFS v3.5.1 and Git 2.46.0, on macOS, FWIW.)  Adding multiple submodules, including several with Git LFS content, seems to work as expected.

You could try upgrading to the latest Git and Git LFS versions, of course.

Beyond that, I'd suggest running the second step (adding the submodule with Git LFS content) with `GIT_TRACE=1`, and perhaps most of all, I would try to build up a reproduction case that's as simple as possible.

For example, if you create a pair of small test repositories in GitHub, one with a single small Git LFS object, can you then reproduce the problem using just those as submodules of another small, new test repository?

Is there anything about your second repository (`https://github.com/vladimirzb/SecundariaORT.git`) which might be an issue?  When you clone it by itself, do you get all the expected contents, without errors?  What do you see if you run `git lfs ls-files` in that clone, for instance?

**Comment by vladimirzb** [comment, marked as answer]
Thank you @chrisd8088 for recommending the use of GIT_TRACE=1. It was instrumental in diagnosing and resolving the issues I was experiencing with Git LFS configuration. Your advice helped me understand the underlying problems and guided me to a successful resolution. I initially encountered an error ('git-lfs filter-process: 1: git-lfs: not found'), and using GIT_TRACE=1 not only showed that Git LFS was not properly set up, but also confirmed that the submodule download and addition processes were actually working, albeit slowly. This was crucial, as I initially thought the process had crashed due to not seeing any immediate feedback. Running git lfs install globally, instead of just within specific repository paths, finally fixed the issue. This is why I could add submodules that didn't have Git LFS, but the one with LFS didn't work because it wasn't found, as I hadn't run git lfs install globally. Thanks for steering me in the right direction!

**Reply by chrisd8088** [reply]
Glad to hear you found the issue, and that makes perfect sense as an explanation.  Sorry I didn't think of it!  But thanks for writing it up, as that likely help someone else in the future.
